### PR TITLE
PIMS-29 Add Project Status History

### PR DESCRIPTION
**Pull request by Fosol**
DB Migration `v01.13.00`

Project Status History will allow us to know the full lifecycle of a project.  This enables better status process control.

I've updated the dependencies I could that didn't seem to break anything.  We'll need to upgrade node and resolve the DevOps dependencies another day.

> I was unable to refactor redux in this PR.  The rabbit hole is just too deep.  I'll see if I can do this on the side of my desk at another point in time.

### CODE REVIEW

**Comment by Fosol** [comment]
@LawrenceLau2020 We'll need to either update your PR with a new migration `v01.13.01`, or we'll need to resolve the questions about your PR and merge it before this PR.

File: backend/dal/Services/Admin/Concrete/AgencyService.cs [hunk 1]
```diff
@@ -110,6 +111,10 @@ public override void Add(Agency agency)
         {
             agency.ThrowIfNull(nameof(agency));
 
+            // Enforce unique constraint on the Code.
+            var isUnique = !this.Context.Agencies.Any(a => a.Code == agency.Code && a.ParentId == agency.ParentId);
```
Comment (Fosol) [inline backend/dal/Services/Admin/Concrete/AgencyService.cs:115]: Fixing bug where an agency with null parent can have the same code.

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 2]
```diff
@@ -163,38 +163,6 @@ public Project Get(int id)
             return project;
         }
 
-        /// <summary>
-        /// Get the project for the specified 'id'.
-        /// </summary>
-        /// <param name="id"></param>
-        /// <param name="includes"></param>
-        /// <returns></returns>
-        public Project Get(int id, params Expression<Func<Project, object>>[] includes)
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:172]: Removed unused function

File: backend/dal/Services/Concrete/ProjectService.cs [hunk 3]
```diff
@@ -539,6 +476,9 @@ public async System.Threading.Tasks.Task<Project> UpdateAsync(Project project)
                 var statusTaskIds = fromStatus.Status.Tasks.Select(t => t.Id);
                 var incompleteStatusTaskIds = incompleteTaskIds.Intersect(statusTaskIds);
                 if (originalProject.Tasks.Any(t => !t.Task.IsOptional && !t.IsCompleted && incompleteStatusTaskIds.Contains(t.TaskId))) throw new InvalidOperationException("Not all required tasks have been completed.");
+
+                // Add a project status history to record the change.
+                this.Context.ProjectStatusHistory.Add(new ProjectStatusHistory(originalProject));
```
Comment (Fosol) [inline backend/dal/Services/Concrete/ProjectService.cs:481]: Save status history

File: frontend/src/components/maps/leaflet/InfoSlideOut/HeaderActions.test.tsx [hunk 4]
```diff
@@ -12,6 +12,18 @@ import { render } from '@testing-library/react';
 import { noop } from 'lodash';
 import HeaderActions from './HeaderActions';
 import { mockParcel } from './InfoContent.test';
+import { useKeycloak } from '@react-keycloak/web';
```
Comment (Fosol) [inline frontend/src/components/maps/leaflet/InfoSlideOut/HeaderActions.test.tsx:15]: With the update to the keycloak package it required adding this to a number of tests.  Ideally I'd have time to clean it up and make it a centralized helper.

File: frontend/src/features/projects/constants/ReviewWorkflowStatus.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
+export enum ReviewWorkflowStatus {
```
Comment (Fosol) [inline frontend/src/features/projects/constants/ReviewWorkflowStatus.ts:1]: Re-organized a bunch of files to improve maintainability.

File: frontend/src/features/projects/spl/forms/SurplusPropertyListForm.tsx [hunk 6]
```diff
@@ -134,12 +132,13 @@ const SurplusPropertyListForm = ({
             Change the Status
           </Form.Label>
           <Form.Group>
-            {values.workflowCode === DisposalWorkflows.Erp && (
+            {values.workflowCode === DisposalWorkflows.Erp && wasInSpl(values) && (
```
Comment (Fosol) [inline frontend/src/features/projects/spl/forms/SurplusPropertyListForm.tsx:135]: We can no make decisions on what to show based on project status history.

**Comment by LawrenceLau2020** [comment]
> DB Migration `v01.13.00`
> 
> Project Status History will allow us to know the full lifecycle of a project. This enables better status process control.
> 
> I've updated the dependencies I could that didn't seem to break anything. We'll need to upgrade node and resolve the DevOps dependencies another day.
> 
> > I was unable to refactor redux in this PR.  The rabbit hole is just too deep.  I'll see if I can do this on the side of my desk at another point in time.

How necessary is it to refactor redux and how long can we delay the refactoring?

**Comment by Fosol** [comment]
We don't need to refactor redux for this PR.  However moving forward it'll be come more difficult to maintain without apply the modern practices I've asked the MOTI team to apply.  The challenge with refactoring presently is that I haven't figured out a way to do it incrementally.  I'm researching a solution that will allow the old code to continue working even if we start using the new structure.

File: backend/dal/Configuration/ProjectStatusHistoryConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            builder.ToTable("ProjectStatusHistory");
+
+            builder.HasKey(m => m.Id);
+            builder.Property(m => m.Id)
+                .IsRequired()
+                .ValueGeneratedOnAdd();
+
+            builder.HasOne(m => m.Project).WithMany(s => s.StatusHistory).HasForeignKey(m => m.ProjectId).OnDelete(DeleteBehavior.Cascade);
+            builder.HasOne(m => m.Status).WithMany().HasForeignKey(m => m.StatusId).OnDelete(DeleteBehavior.Restrict);
+            builder.HasOne(m => m.Workflow).WithMany().HasForeignKey(m => m.WorkflowId).OnDelete(DeleteBehavior.Restrict);
```
Comment (LawrenceLau2020) [inline backend/dal/Configuration/ProjectStatusHistoryConfiguration.cs:24]: Is there a way to delete a project, and if so does it only delete all associated projectstatushistory records as well as the project itself?

**Review by LawrenceLau2020 — APPROVED** [review]
Do all existing projects get assigned a projectStatusHistory record with migration or would a projectStatusHistory only get created when a project gets saved?

**Pull request merged**